### PR TITLE
Feat/daily quests

### PR DESCRIPTION
**Pull request by anhnh12**
**What does this PR do?**
- [x] For now we support 6 quests: `gm`, `vote`, `ticker`, `watchlist`, `gift`, `trade`
- [x] New quests APIs
  1. Get user's quests list
  2. Claim rewards

**Bonus**: all kinds of quest progress are updated internally in BE side, except for `trade`
-> Reason: `$trade` command do not make any request to mochi-api so we need to expose one more API to allow mochi-bot directly update `trade` quest progress

### CODE REVIEW

File: pkg/handler/quest.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// @Accept      json
+// @Produce     json
+// @Param       req query request.GetUserQuestListRequest true "request"
+// @Success     200 {object} response.GetUserQuestListResponse
+// @Router      /quests [get]
+func (h *Handler) GetUserQuestList(c *gin.Context) {
+	var req request.GetUserQuestListRequest
+	if err := c.ShouldBindQuery(&req); err != nil {
+		h.log.Error(err, "[handler.GetUserQuestList] ShouldBindQuery() failed")
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
```
Comment (trkhoi) [inline pkg/handler/quest.go:25]: c.JSON(http.StatusBadRequest, response.CreateResponse[any](nil, nil, err, nil))

File: pkg/handler/quest.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// @Accept      json
+// @Produce     json
+// @Param       req body request.ClaimQuestsRewardsRequest true "request"
+// @Success     200 {object} response.ClaimQuestsRewardsResponse
+// @Router      /quests/claim [POST]
+func (h *Handler) ClaimQuestsRewards(c *gin.Context) {
+	var req request.ClaimQuestsRewardsRequest
+	if err := c.BindJSON(&req); err != nil {
+		h.log.Error(err, "[handler.ClaimQuestReward] BindJSON() failed")
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
```
Comment (trkhoi) [inline pkg/handler/quest.go:50]: c.JSON(http.StatusBadRequest, response.CreateResponse[any](nil, nil, err, nil))

File: pkg/handler/quest.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		return
+	}
+
+	data, err := h.entities.ClaimQuestsRewards(req)
+	if err != nil {
+		h.log.Error(err, "[handler.ClaimQuestsRewards] entity.ClaimQuestsRewards() failed")
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
```
Comment (trkhoi) [inline pkg/handler/quest.go:57]: c.JSON(http.StatusBadRequest, response.CreateResponse[any](nil, nil, err, nil))

File: pkg/handler/quest.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"net/http"
+
+	"github.com/defipod/mochi/pkg/logger"
+	"github.com/defipod/mochi/pkg/request"
+	"github.com/defipod/mochi/pkg/response"
+	"github.com/gin-gonic/gin"
+)
+
+// GetUserQuestList     godoc
```
Comment (trkhoi) [inline pkg/handler/quest.go:12]: make gen-swagger

**Review by trkhoi — CHANGES_REQUESTED** [review]

File: pkg/entities/webhook.go [hunk 5]
```diff
@@ -464,6 +496,15 @@ func (e *Entity) WebhookUpvoteStreak(userID, source string) error {
...
 
+	// handle quest logs
+	log := &model.QuestUserLog{
+		UserID: userID,
+		Action: model.QuestAction(model.VOTE),
+	}
+	if err := e.UpdateUserQuestProgress(log); err != nil {
+		e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.WebhookUpvoteStreak] entity.UpdateUserQuestProgress() failed")
+	}
+
```
Comment (trkhoi) [inline pkg/entities/webhook.go:507]: should we return here

File: pkg/entities/webhook.go [hunk 6]
```diff
@@ -443,6 +467,14 @@ func (e *Entity) WebhookUpvoteStreak(userID, source string) error {
...
 			return fmt.Errorf("failed to create new user upvote streak: %v", err)
 		}
+		// handle quest logs
+		log := &model.QuestUserLog{
+			UserID: userID,
+			Action: model.QuestAction(model.VOTE),
+		}
+		if err := e.UpdateUserQuestProgress(log); err != nil {
+			e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.WebhookUpvoteStreak] entity.UpdateUserQuestProgress() failed")
+		}
```
Comment (trkhoi) [inline pkg/entities/webhook.go:477]: should we return here

File: pkg/entities/webhook.go [hunk 7]
```diff
@@ -241,11 +254,22 @@ func (e *Entity) newUserGM(authorAvatar, authorUsername, discordID, guildID, cha
...
 	}
 
+	// handle quest logs
+	log := &model.QuestUserLog{
+		GuildID: &guildID,
+		UserID:  discordID,
+		Action:  model.QuestAction(model.GM),
+	}
+	if err := e.UpdateUserQuestProgress(log); err != nil {
+		e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.newUserGM] entity.UpdateUserQuestProgress() failed")
```
Comment (trkhoi) [inline pkg/entities/webhook.go:269]: should we return here

File: pkg/entities/webhook.go [hunk 8]
```diff
@@ -197,7 +198,18 @@ func (e *Entity) newUserGM(authorAvatar, authorUsername, discordID, guildID, cha
...
+		}
+
+		// handle quest logs
+		log := &model.QuestUserLog{
+			GuildID: &guildID,
+			UserID:  discordID,
+			Action:  model.QuestAction(model.GM),
+		}
+		if err := e.UpdateUserQuestProgress(log); err != nil {
+			e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.newUserGM] entity.UpdateUserQuestProgress() failed")
```
Comment (trkhoi) [inline pkg/entities/webhook.go:212]: should we return

File: pkg/entities/user.go [hunk 9]
```diff
@@ -421,9 +421,18 @@ func (e *Entity) SendGiftXp(req request.GiftXPRequest) (*response.HandleUserActi
...
 	}
+	// handle quest logs
+	log := &model.QuestUserLog{
+		UserID:  req.UserDiscordID,
+		GuildID: &req.GuildID,
+		Action:  model.QuestAction(model.GIFT),
+	}
+	if err := e.UpdateUserQuestProgress(log); err != nil {
+		e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.SendGiftXp] entity.UpdateUserQuestProgress() failed")
+	}
```
Comment (trkhoi) [inline pkg/entities/user.go:435]: should we return

**Review by trkhoi — CHANGES_REQUESTED** [review]

File: pkg/entities/defi.go [hunk 10]
```diff
@@ -35,7 +35,14 @@ func (e *Entity) GetHistoricalMarketChart(c *gin.Context) (*response.CoinPriceHi
...
 	}
-
+	// handle quest logs
+	log := &model.QuestUserLog{
+		UserID: req.DiscordID,
+		Action: model.QuestAction(model.TICKER),
+	}
+	if err := e.UpdateUserQuestProgress(log); err != nil {
+		e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.GetHistoricalMarketChart] entity.UpdateUserQuestProgress() failed")
+	}
```
Comment (trkhoi) [inline pkg/entities/defi.go:45]: should we return

File: pkg/entities/defi.go [hunk 11]
```diff
@@ -615,6 +622,14 @@ func (e *Entity) GetUserWatchlist(req request.GetUserWatchlistRequest) (*[]respo
...
 		data = append(data, item)
 	}
+	// handle quest logs
+	log := &model.QuestUserLog{
+		UserID: req.UserID,
+		Action: model.QuestAction(model.WATCHLIST),
+	}
+	if err := e.UpdateUserQuestProgress(log); err != nil {
+		e.log.Fields(logger.Fields{"log": log}).Error(err, "[entity.GetUserWatchlist] entity.UpdateUserQuestProgress() failed")
+	}
```
Comment (trkhoi) [inline pkg/entities/defi.go:632]: should we return

File: pkg/handler/quest.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+	data, err := h.entities.GetUserQuestList(req)
+	if err != nil {
+		h.log.Fields(logger.Fields{"req": req}).Error(err, "[handler.GetUserQuestList] entity.GetUserQuestList() failed")
+		c.JSON(http.StatusInternalServerError, response.CreateResponse[any](nil, nil, err, nil))
+		return
+	}
+	c.JSON(http.StatusOK, &response.GetUserQuestListResponse{Data: data})
```
Comment (nguyennh4522) [inline pkg/handler/quest.go:34]:
```suggestion
	c.JSON(http.StatusOK, response.CreateResponse(data, nil, err, nil))
```

**Review by trkhoi — CHANGES_REQUESTED** [review]

File: migrations/schemas/20221011152829-quests.sql [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	target INT NOT NULL DEFAULT 0,
+	created_at timestamptz DEFAULT now(),
+	FOREIGN KEY (guild_id) REFERENCES discord_guilds(id)
+);
+
+CREATE TABLE IF NOT EXISTS quests_user_list (
+	id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+	user_id TEXT NOT NULL,
+	quest_id UUID NOT NULL,
+	action quest_action NOT NULL,
```
Comment (nguyennh4522) [inline migrations/schemas/20221011152829-quests.sql:45]: maybe you miss default value for `action`

File: pkg/entities/webhook.go (hunk 6 above)
Comment (anhnh12) [reply pkg/entities/webhook.go:477]: return err here may prevent below functions from execution
this quest handler should not affect original function flow imo

File: pkg/entities/webhook.go (hunk 7 above)
Comment (anhnh12) [reply pkg/entities/webhook.go:269]: same as above

File: pkg/entities/webhook.go (hunk 8 above)
Comment (anhnh12) [reply pkg/entities/webhook.go:212]: same as above

File: pkg/entities/webhook.go (hunk 5 above)
Comment (anhnh12) [reply pkg/entities/webhook.go:507]: return err here may prevent below functions from execution
this quest handler should not affect original function flow imo

File: pkg/entities/defi.go (hunk 10 above)
Comment (anhnh12) [reply pkg/entities/defi.go:45]: same as above

File: pkg/entities/user.go (hunk 9 above)
Comment (anhnh12) [reply pkg/entities/user.go:435]: same as above

File: pkg/entities/webhook.go (hunk 6 above)
Comment (anhnh12) [reply pkg/entities/webhook.go:477]: same as above

File: pkg/entities/defi.go (hunk 11 above)
Comment (anhnh12) [reply pkg/entities/defi.go:632]: same as above

File: migrations/schemas/20221011152829-quests.sql (hunk 13 above)
Comment (anhnh12) [reply migrations/schemas/20221011152829-quests.sql:45]: this should be passed actively, so no default value imo

**Review by trkhoi — APPROVED** [review]

**Review by nguyennh4522 — APPROVED** [review]

**Pull request merged**